lib/srv/db/spanner/grpcserver.go

```diff
@@ -0,0 +1,467 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package spanner
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	spannerapi "cloud.google.com/go/spanner/apiv1"
+	"cloud.google.com/go/spanner/apiv1/spannerpb"
+	"github.com/gravitational/trace"
+	"google.golang.org/api/option"
+	gtransport "google.golang.org/api/transport/grpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/srv/db/common/role"
+)
+
+type upstream[T any] interface {
+	Recv() (T, error)
+	grpc.ClientStream
+}
+
+type downstream[T any] interface {
+	Send(T) error
+	grpc.ServerStream
+}
+
+func proxyServerStream[T any](up upstream[T], down downstream[T]) error {
+	for {
+		res, err := up.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// EOF is expected and signals end of the server stream.
+				return nil
+			}
+			return trace.Wrap(err)
+		}
+		if err := down.Send(res); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+func (e *Engine) CreateSession(ctx context.Context, req *spannerpb.CreateSessionRequest) (*spannerpb.Session, error) {
+	if err := e.authorizeRPCRequest(ctx, req.GetDatabase(), "CreateSession", req, false); err != nil {
```

Comment:
What is the difference between database, session name and session ?

```
e.authorizeRPCRequest(ctx, req.GetDatabase()
e.authorizeRPCRequest(ctx, req.GetName()
e.authorizeRPCRequest(ctx, req.GetSession()
```

this properties are forwarded to  [parseDatabaseName](https://github.com/gravitational/teleport/blob/fce5a63d39a5acd82a595910969e31fe3b52be80/lib/srv/db/spanner/grpcserver.go#L449) as targetID and used as `db_name` in our RBAC. 


Are you sure taht  req.GetSession() will return string with proper format  ?

Reply:
project id: `projects/<project>`
instance id: `<project-id>/instances/<instance>`
database id: `<instance-id>/databases/<database>`
session id: `<database-id>/sessions/<session>`

So a session id looks like: `projects/<project>/instances/<instance>/databases/<database>/sessions/<session>`

When creating a session, as in `CreateSession`, the caller does not use a session ID, they send a database ID.
All RPCs have either a database ID or a session ID in them.

I will clarify this better with comments for the code.

> Are you sure taht req.GetSession() will return string with proper format ?

Clients that are not trying to do anything malicious will send an RPC with the proper format ;)
Anything that does not return the proper format will be denied.
`checkAccess` ensures that `targetID` is prefixed by the correct instance id and then parses and extracts the <database> .
Any failure to parse is access denied.
Even it were not denied, it wouldn't parse on Google's end either, so the request would fail even if we did forward it.

Reply:
clarifying godoc added in 3e03af22dd42e0eef13f8a239f8814ea79bc869e